### PR TITLE
Refactor FXIOS-15205 [Bookmarks Panel] Fix bugs with the books panel search causing crashes on deleted bookmarks in search results

### PR DESCRIPTION
**Pull request by ih-codes**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-15205)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/32719)

## :bulb: Description

Fix issues with the bookmarks panel search not accessing the correct underlying arrays, which would lead to runtime crashes when deleting bookmarks turned up in search results. Streamline some empty state/bottom bar updates.

We will need to merge #32823 first.

### Testing Instructions
ℹ️  To test this new feature, please turn the `bookmarks-search-feature` to `true` in the nimbus feature yaml (no debug toggle yet, sorry).
Please also test with the feature disabled (`false`) to ensure nothing has broken in regular use as well.

Sample of test edge cases:
- Search bookmarks in a folder, then attempt to delete one of the search results (via press and hold context menu)
- Enter table editing mode using "Edit" and drill down into editing a bookmark, or a folder, and pop back up
- Entering "Edit" mode should disable "Search" mode, and vice versa
- Test searching bookmarks from desktop sync
    - Search should recursively search bookmarks by URL and Title at all folder depths, but the Desktop sync folders are separate from Mobile (current limitation in this implementation...)
- Bookmark pin / unpin via the long press context menu works as expected 
- Empty bookmarks folders should not show a search option, but a bookmark folder that just contains folders should show the search option

## :pencil: Checklist
- [x] I filled in the ticket numbers and a description of my work
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] I ensured unit tests pass and wrote tests for new code
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If adding or modifying strings, I read the [guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/How-to-add-and-modify-Strings) and will request a string review from l10n
- [ ] If needed, I updated documentation and added comments to complex code

### CODE REVIEW

File: firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift [hunk 1]
```diff
@@ -242,10 +240,9 @@ final class BookmarksViewController: SiteTableViewController,
     override func viewDidDisappear(_ animated: Bool) {
         super.viewDidDisappear(animated)
 
+        // If needed, exit the search when the view disappears
         if state == .bookmarks(state: .search) {
             exitSearchState()
-            let substate: LibraryPanelSubState = viewModel.isRootNode ? .mainView : .inFolder
-            updatePanelState(newState: .bookmarks(state: substate))
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift:248]: FYI this code has been moved into `exitSearchState`

File: firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift [hunk 2]
```diff
@@ -850,9 +859,6 @@ extension BookmarksViewController {
         // When the Done button is tapped and search is active, exit search mode
         if state == .bookmarks(state: .search) {
             exitSearchState()
-            updatePanelState(newState: viewModel.isRootNode
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift:853]: This is handled in `exitSearchState`

File: firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift [hunk 3]
```diff
@@ -903,8 +909,6 @@ extension BookmarksViewController: UISearchBarDelegate {
 
     func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
         exitSearchState()
-        let substate: LibraryPanelSubState = viewModel.isRootNode ? .mainView : .inFolder
-        updatePanelState(newState: .bookmarks(state: substate))
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift:907]: This is handled in `exitSearchState`

File: firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift [hunk 4]
```diff
@@ -599,18 +608,18 @@ final class BookmarksViewController: SiteTableViewController,
     }
 
     func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
-        guard !viewModel.isSearching else { return false }
+        guard state != .bookmarks(state: .search) else { return false }
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift:616]: The viewModel's `isSearching` (renamed `isShowingSearchResults`) actually is true just when the results should be replaced with the filtered search results in the tableView. 

When the search bar is emptied of text (e.g. with the clear button, or delete on keyboard, etc.), then `isShowingSearchResults` becomes false and the current folder is shown instead.

File: firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift [hunk 5]
```diff
@@ -520,20 +523,26 @@ final class BookmarksViewController: SiteTableViewController,
...
-               !viewModel.isSearching,
+               state != .bookmarks(state: .search),
                isCurrentFolderEditable(at: indexPath) {
                 // Only show detail controller for editable nodes
                 bookmarkCoordinatorDelegate?.showBookmarkDetail(for: node, folder: bookmarkFolder)
             }
             return
         }
 
-        updatePanelState(newState: .bookmarks(state: .inFolder))
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift:531]: I'm actually pretty sure this was not intentional since we push a new panel to the nav stack when drilling down. THIS current view's state should not change. 🤔

File: firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift [hunk 6]
```diff
@@ -728,19 +742,17 @@ final class BookmarksViewController: SiteTableViewController,
...
-                if let site {
-                    self.presentContextMenu(for: site, with: indexPath, completionHandler: {
-                        return self.contextMenu(for: site, with: indexPath)
-                    })
-                } else if let bookmarkNode = self.viewModel.displayedBookmarkNodes[safe: indexPath.row],
-                          bookmarkNode.type == .folder,
-                          self.isCurrentFolderEditable(at: indexPath) {
-                    self.presentContextMenu(for: bookmarkNode, indexPath: indexPath)
-                }
+        guard let bookmark = viewModel.displayedBookmarkNodes[safe: indexPath.row] else { return }
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Library/Bookmarks/BookmarksViewController.swift:745]: Just a tweak to get unit tests passing so I don't have to also add a bunch of mocking in this PR 😅

**Comment by Foxbolts** [comment]
You can search from anywhere in the bookmarks file structure, but search seems to only search subfolders recursively, as opposed to the whole data stucture (is this intentional? I could go either way), but when editing a bookmark during search, if that bookmark is not parented to the current folder, it will re-parent immediately upon entering the edit form (you'll notice the "Save in" field will automatically change to the current folder), causing a change in file structure, even if no edits are actually made.

Not sure if it is caused by your PR per se, so feel free to file future work.

https://github.com/user-attachments/assets/8c8bf664-5ef7-41e3-8932-091bfb6abadc

**Comment by Foxbolts** [comment]
Edit 2: Ignore this one

Deleting bookmarks from a subdirectory of the current directory in search seems to always cause a crash. Furthermore, if you try to delete an item from a subdirectory and it's position in the search results screen is the same as a non-empty folder in the current folder, it prompts the non-empty folder deletion confirmation alert

Again, feel free to file this separately if unrelated 🙂

Edit: Actually, deleting any bookmark whose index differs from its original position in the search results screen seems to result in a crash

https://github.com/user-attachments/assets/84d81642-0370-48e7-8768-43c0038f60d6

**Comment by Foxbolts** [comment]
Edit: Ignore this one too

After deleting the last saved bookmark, once the empty state appears, you cannot close out of search mode unless you dismiss the sheet or switch library panels

Same disclaimer as the other bugs 😆 

https://github.com/user-attachments/assets/1fa2065b-c80c-4bc3-bad3-d3109698f8a4

**Comment by Foxbolts** [comment]
Deleting an empty folder wrongly prompts the non-empty folder deletion confirmation alert when trying to delete the folder after deleting the last saved bookmark in that folder via the search results screen

https://github.com/user-attachments/assets/19af461e-feef-4231-af53-740fdb9d4c87

**Comment by ih-codes** [comment]
> You can search from anywhere in the bookmarks file structure, but search seems to only search subfolders recursively, as opposed to the whole data stucture (is this intentional? I could go either way),

I should honestly check with Andy about this. That is "intentional" in the sense the contributor emulated the history search, but history doesn't have multi-level folders, so it's not a one-to-one comparison. 🤔  Might be better to treat this more like a "filter" than a "search" if we're leaving it in the current implementation...

> but when editing a bookmark during search, if that bookmark is not parented to the current folder, it will re-parent immediately upon entering the edit form (you'll notice the "Save in" field will automatically change to the current folder), causing a change in file structure, even if no edits are actually made.
> 
> Not sure if it is caused by your PR per se, so feel free to file future work.
> Simulator.Screen.Recording.-.iPhone.17.-.2026-03-30.at.19.37.54.mov

Aha, this is why. Probably non-trivial to implement so we can get the correct parent, so I might move this into a separate bug ticket. 😔 

Screenshot 2026-03-31 at 11 29 12 AM

Filed a bug ticket here: https://mozilla-hub.atlassian.net/browse/FXIOS-15291

**Comment by ih-codes** [comment]
> Deleting an empty folder wrongly prompts the non-empty folder deletion confirmation alert when trying to delete the folder after deleting the last saved bookmark in that folder via the search results screen

Nice catch on that, totally wouldn't have noticed this myself!!

Took a brief look. Seems like we need to reload the bookmarks at the very least to refresh some local folder data after deleting a bookmark during search. 🤔 I've filed a separate bug ticket to track it: FXIOS-15296 / #32845

**Comment by ih-codes** [comment]
Will merge this once https://github.com/mozilla-mobile/firefox-ios/pull/32823 is in.
